terminal/command: Add 'reload' command

alexsaezm opened the pull request:
These changes add the 'reload' command, which allows us to rebuild the project
and start the debugging session again. Currently, if the project's code is
updated while debugging it, Delve shows the new source code, but it's still
running the old one. With 'reload', the whole binary is rebuilt, and the
process starts again.

Fixes #1551

aarzilli commented on service/debugger/debugger.go:
```diff
@@ -420,6 +428,12 @@ func (d *Debugger) Restart(rerecord bool, pos string, resetArgs bool, newArgs []
 		p, err = d.recordingRun(run)
 		d.recordingDone()
 	} else {
+		if rebuild {
```
This should go before the `if recorded` line, otherwise it won't have an effect on recorded targets.

Also, breakpoints that can't be set by (file, line) should be discarded in the `for _, oldBp := ...` loop below.

aarzilli commented on service/debugger/debugger.go:
```diff
@@ -420,6 +428,12 @@ func (d *Debugger) Restart(rerecord bool, pos string, resetArgs bool, newArgs []
 		d.recordingDone()
 	} else {
+		if rebuild {
+			err := gobuild.GoBuild(d.processArgs[0], d.config.Packages, d.config.BuildFlags)
```
Some error message should be returned if the debugger was started with `attach` or `exec`, also this doesn't work if the debugger was started with `test`.

aarzilli commented on pkg/terminal/command.go:
```diff
@@ -142,6 +142,7 @@ For live targets the command takes the following forms:
 If newargv is omitted the process is restarted (or re-recorded) with the same argument vector.
 If -noargs is specified instead, the argument vector is cleared.
 `},
+		{aliases: []string{"reload"}, group: runCmds, cmdFn: c.reload, allowedPrefixes: revPrefix, helpMsg: "Refresh code."},
```
I think `rebuild` would be a better name.

derekparker commented on cmd/dlv/cmds/commands.go:
```diff
@@ -47,6 +47,10 @@ var (
 	// BuildFlags is the flags passed during compiler invocation.
 	BuildFlags string
+	// DlvArgs DlvArgs contains debug arguments.
+	DlvArgs []string
```
I don't think we need to add these here and export them.

derekparker commented on cmd/dlv/cmds/commands.go:
```diff
@@ -477,20 +481,20 @@ func traceCmd(cmd *cobra.Command, args []string) {
 			}
 
-			processArgs = append([]string{debugname}, targetArgs...)
+			processArgs = append([]string{debugname}, TargetArgs...)
```
(I can't add a review comment on the appropriate line cause it wasn't changed so I'll add it here)

You must fill in the new struct members you added to `service.Config` below (`Packages` and `BuildFlags`). Currently they are not being passed through and are always empty, it appears.

alexsaezm commented:
I added a test for the feature, and also move the `ExecuteKind` from `cmds` to `debugger`. Also, I renamed the feature because yeah, makes total sense to call it rebuild.

aarzilli commented on service/debugger/debugger.go:
```diff
@@ -434,6 +434,15 @@ func (d *Debugger) Restart(rerecord bool, pos string, resetArgs bool, newArgs []
 	var err error
+
+	// TODO executeKind from command.go
+	if rebuild && d.canRestart() {
```
The code above has a `if !d.canRestart() { return ... }` why do you need to check this here too?

Also the loop that sets breakpoints below should discard the ones that can't be set by file/line.

alexsaezm replied:
You are totally right. I removed the d.canRestart.

Can you please clarify a little more the loop part? I don't really understand what you mean.

aarzilli replied:
In this function there's a loop, `for _, oldBp := ...` that restores breakpoints. It either does it by file:line (`if len(oldBp).File > 0`) or by address (`else`). We shouldn't try to set them by address because addresses have changed after rebuilding.

alexsaezm replied:
Thanks for the explanation! I updated the loop to avoid the case

aarzilli commented on go.mod:
```diff
@@ -20,6 +20,6 @@ require (
 	golang.org/x/arch v0.0.0-20190927153633-4e8777c89be4
 	golang.org/x/sys v0.0.0-20190626221950-04f50cda93cb
-	golang.org/x/tools v0.0.0-20191127201027-ecd32218bd7f
+	golang.org/x/tools v0.0.0-20200415034506-5d8e1897c761
```
I don't think this needs an updated version of golang.org/x/tools so this should be reverted. If I'm wrong and it does the vendor directory should also be updated.

alexsaezm replied:
Yeah, for sure it is not necessary. My bad, I think I update this while doing the rebase of the branch for some reason.

aarzilli commented on service/debugger/debugger.go:
```diff
@@ -420,6 +439,20 @@ func (d *Debugger) Restart(rerecord bool, pos string, resetArgs bool, newArgs []
+		// d.config.Kind is different from ExecutingGeneratedFile that means we
+		// didn't build the binary and we are just running it
+		if d.config.Kind == ExecutingExistingFile {
+			return nil, fmt.Errorf("cannot restart process Delve did not create")
```
s/cannot restart/cannot rebuild/

alexsaezm replied:
Fixed

aarzilli commented on cmd/dlv/cmds/commands.go:
```diff
@@ -387,11 +387,11 @@ func dapCmd(cmd *cobra.Command, args []string) {
 		}
-		dlvArgs, targetArgs := splitArgs(cmd, args)
-		if len(dlvArgs) > 0 {
+		DlvArgs, TargetArgs := splitArgs(cmd, args)
```
why the capitalization?

alexsaezm replied:
Because I totally mixed that up :disappointed: . Changed

aarzilli commented on service/debugger/debugger.go:
```diff
@@ -420,6 +439,20 @@ func (d *Debugger) Restart(rerecord bool, pos string, resetArgs bool, newArgs []
+		if d.config.Kind == ExecutingExistingFile {
+			return nil, fmt.Errorf("cannot restart process Delve did not create")
+		}
+		err := gobuild.GoBuild(d.processArgs[0], d.config.Packages, d.config.BuildFlags)
```
This should call GoTestBuild if `Kind == ExecutingGeneratedTest`.

aarzilli commented on service/test/integration2_test.go:
```diff
@@ -74,6 +75,9 @@ func startServer(name string, buildFlags protest.BuildFlags, t *testing.T) (clie
 			CheckGoVersion: true,
+			Packages:       []string{fixture.Source},
+			BuildFlags:     "", // build flags can be an empty string here because the only test that uses it, does not set special flags.
+			Kind:           debugger.ExecutingGeneratedTest,
```
This should say ExecutingGeneratedFile, it doesn't really matter but for consistency that's what they are.

alexsaezm replied:
Makes sense. I changed it.

aarzilli commented on service/test/integration2_test.go:
```diff
@@ -110,6 +114,7 @@ func TestRunWithInvalidPath(t *testing.T) {
 		APIVersion:  2,
 		Debugger: debugger.Config{
 			Backend: testBackend,
+			Kind:    debugger.ExecutingGeneratedTest,
```
Also ExecutingGeneratedFile.

alexsaezm commented on service/rpccommon/server.go:
```diff
@@ -116,6 +116,8 @@ func (s *ServerImpl) Run() error {
 		Foreground:           s.config.Foreground,
 		DebugInfoDirectories: s.config.DebugInfoDirectories,
 		CheckGoVersion:       s.config.CheckGoVersion,
+		Packages:             s.config.Packages,
```
It should work now :)

aarzilli commented on service/debugger/debugger.go:
```diff
@@ -420,6 +439,21 @@ func (d *Debugger) Restart(rerecord bool, pos string, resetArgs bool, newArgs []
+		// d.config.Kind is different from ExecutingGeneratedFile that means we
+		// didn't build the binary and we are just running it
+		if d.config.Kind == ExecutingExistingFile {
+			err = gobuild.GoTestBuild(d.processArgs[0], d.config.Packages, d.config.BuildFlags)
```
This is still wrong:

ExecutingExistingFile, ExecutingOther → it can't be rebuilt at all (either an separately built executable file or a core dump/recording)
ExecutingGeneratedFile → it can be rebuilt by GoBuild
ExecutingGeneratedTest → it can be rebuilt by GoTestBuild

it doesn't make sense to call GoTestBuild on ExecutingExistingFile and it doesn't make sense to call GoBuild on anything other than ExecutingGeneratedFile.

alexsaezm replied:
Updated. Thanks for the explanation :)

aarzilli commented:
I'm sorry I forgot to tell you this yesterday but there is a conflict between your PR and something that has already been merged into master that needs to be solved. You need to do a git rebase.

alexsaezm commented:
I rebased the fork to keep it up to date with master.

derekparker commented on pkg/terminal/command.go:
```diff
@@ -143,6 +143,7 @@ For live targets the command takes the following forms:
 If newargv is omitted the process is restarted (or re-recorded) with the same argument vector.
 If -noargs is specified instead, the argument vector is cleared.
 `},
+		{aliases: []string{"rebuild"}, group: runCmds, cmdFn: c.rebuild, allowedPrefixes: revPrefix, helpMsg: "Rebuild the code."},
```
We need more documentation here. Add in more information about how it will recompile code, etc, etc...

derekparker commented on pkg/terminal/command.go:
```diff
@@ -1073,6 +1074,18 @@ func printcontextNoState(t *Term) {
+	defer t.onStop()
+	discarded, err := t.client.Restart(true)
+	if len(discarded) > 0 {
+		fmt.Printf("Some of the breakpoints could not be recovered")
```
nit: no need for capitalization. Also, maybe rephrase to `not all breakpoints could be restored`.

derekparker commented on service/config.go:
```diff
@@ -36,4 +36,17 @@ type Config struct {
+
+	// TTY is passed along to the target process on creation. Used to specify a
+	// TTY for that process.
+	TTY string
```
This flag can be removed.

derekparker commented on service/debugger/debugger.go:
```diff
@@ -421,6 +440,26 @@ func (d *Debugger) Restart(rerecord bool, pos string, resetArgs bool, newArgs []
+			}
+		default:
+			// We cannot build a process that we didn't start, because we don't know
+			// how it was build.
```
s/build/built/

derekparker commented on service/debugger/debugger.go:
```diff
@@ -450,6 +489,10 @@ func (d *Debugger) Restart(rerecord bool, pos string, resetArgs bool, newArgs []
 			}
 			createLogicalBreakpoint(p, addrs, oldBp)
 		} else {
+			// Avoid setting a breakpoint based on address when rebulding
```
s/rebulding/rebuilding/

aarzilli commented:
> @aarzilli did you have a chance to take another look as well?

I have now. I don't have anything to add. For the rebuild command documentation I think something like this would work:

```
Rebuild the target executable and restarts it. Does not work if the executable was not built by delve.
```

derekparker commented on cmd/dlv/cmds/commands.go:
```diff
@@ -465,7 +465,7 @@ func traceCmd(cmd *cobra.Command, args []string) {
 		var processArgs []string
 
-		dlvArgs, targetArgs := splitArgs(cmd, args)
+		dlvArgs, TargetArgs := splitArgs(cmd, args)
```
This does not need to be exported.

derekparker commented on service/config.go:
```diff
@@ -36,4 +36,13 @@ type Config struct {
 	// DisconnectChan will be closed by the server when the client disconnects
 	DisconnectChan chan<- struct{}
+
+	// Packages contains the packages that we are debugging.
```
These three fields can be removed because we've added them to the `debugger.Config` as well, which we include within this config already anyway.

derekparker commented on service/debugger/debugger.go:
```diff
@@ -98,6 +108,15 @@ type Config struct {
+	BuildFlags string
+
+	// ExecuteKind contains the kind of the executed program.
+	Kind ExecuteKind
```
nit: `Kind` is a little vague, maybe we stutter a bit and name this field `ExecuteKind` as well.

alexsaezm replied:
Done. `ExecuteKind` sounds perfect. I updated a few more places where it appears.
